components/cop_codegen/src/rpn_function.rs

```diff
@@ -1,49 +1,321 @@
+        let et = input.parse::<RpnFnEvaluableType>()?;
+        Ok(Self {
+            _pat: pat,
+            eval_type: et.eval_type,
+        })
+    }
+}
+
+/// Parses a function signature parameter like `val: &[&Option<T>]`.
+struct VargsRpnFnSignatureParam {
+    _pat: Pat,
+    eval_type: TypePath,
+}
+
+impl parse::Parse for VargsRpnFnSignatureParam {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        let pat = input.parse::<Pat>()?;
+        input.parse::<Token![:]>()?;
+        input.parse::<Token![&]>()?;
+        let slice_inner;
+        bracketed!(slice_inner in input);
+        slice_inner.parse::<Token![&]>()?;
+        let et = slice_inner.parse::<RpnFnEvaluableType>()?;
+        Ok(Self {
+            _pat: pat,
+            eval_type: et.eval_type,
+        })
+    }
+}
+
+/// Parses a function signature return type like `Result<Option<T>>`.
+struct RpnFnSignatureReturnType {
+    eval_type: TypePath,
+}
+
+impl parse::Parse for RpnFnSignatureReturnType {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        input.parse::<Token![->]>()?;
+        let tp = input.parse::<Type>()?;
+        if let Type::Path(TypePath {
+            path: Path { segments, .. },
+            ..
+        }) = &tp
+        {
+            let result_type = (*segments.last().unwrap().value()).clone();
+            if let PathArguments::AngleBracketed(AngleBracketedGenericArguments { args, .. }) =
+                result_type.arguments
+            {
+                let et = parse2::<RpnFnEvaluableType>(args.into_token_stream())?;
+                Ok(Self {
+                    eval_type: et.eval_type,
+                })
+            } else {
+                return Err(Error::new_spanned(
+                    tp,
+                    "expect angle bracketed path arguments",
+                ));
+            }
+        } else {
+            return Err(Error::new_spanned(tp, "expect path"));
+        }
+    }
+}
+
+pub struct RpnFn;
+
+impl RpnFn {
+    pub fn transform(attr: TokenStream, item_fn: TokenStream) -> Result<TokenStream> {
+        let attr = parse2::<RpnFnAttr>(attr)?;
+        let item_fn = parse2::<ItemFn>(item_fn)?;
+
+        // FIXME: The macro cannot handle lifetime definitions now
+        if let Some(lifetime) = item_fn.decl.generics.lifetimes().next() {
+            return Err(Error::new_spanned(
+                lifetime,
+                "Lifetime definition is not allowed",
+            ));
+        }
+
+        if attr.is_varg {
+            Ok(VargsRpnFn::new(attr, item_fn)?.generate())
+        } else {
+            Ok(NormalRpnFn::new(attr, item_fn)?.generate())
+        }
+    }
+}
+
+#[derive(Debug)]
+struct VargsRpnFn {
+    captures: Vec<Expr>,
+    item_fn: ItemFn,
+    arg_type: TypePath,
+    ret_type: TypePath,
+}
+
+impl VargsRpnFn {
+    pub fn new(attr: RpnFnAttr, item_fn: ItemFn) -> Result<Self> {
+        if item_fn.decl.inputs.len() != attr.captures.len() + 1 {
```

Comment:
Are we assuming that vargs functions only have arguments of the same type? Will there be functions like `foo(Int, Real, &[Decimal])`? Perhaps we can unify normal and vargs functions.

Reply:
Yes. I took a look at existing varg functions and they accept same argument types. We can add support for multiple argument type varg function later if they exist later. The change should be pretty simple, the signature is going to be `&[ScalarValueRef]` and there will be no performance difference. However I didn't choose this signature since this kind of flexibility is unnecessary, while on the other hand a signature like `&[&Option<T>]` looks to be very clear.